src/cast_core/src/CrudActor.cpp

```diff
@@ -1132,6 +1135,86 @@ std::string getDbName(const PhaseContext& phaseContext) {
     return phaseDb ? *phaseDb : *actorDb;
 }
 
+enum Units { NANOSECOND, MICROSECOND, MILLISECOND, SECOND, MINUTE, HOUR };
+// represent a random delay
+class Delay {
+public:
+    Delay(const Node& node, GeneratorArgs args)
+        : number_generator{doubleGenerator(node["^TimeSpec"]["value"], args)} {
+        auto unit_string = node["^TimeSpec"]["units"].maybe<std::string>().value_or("seconds");
+
+        // Use string::find here so plurals get parsed correctly.
+        if (unit_string.find("nanosecond") == 0) {
+            units = NANOSECOND;
+        } else if (unit_string.find("microsecond") == 0) {
+            units = MICROSECOND;
+        } else if (unit_string.find("millisecond") == 0) {
+            units = MILLISECOND;
+        } else if (unit_string.find("second") == 0) {
+            units = SECOND;
+        } else if (unit_string.find("minute") == 0) {
+            units = MINUTE;
+        } else if (unit_string.find("hour") == 0) {
+            units = HOUR;
+        } else {
+            std::stringstream msg;
+            msg << "Invalid unit: " << unit_string << " for genny::TimeSpec field in config";
+            throw genny::InvalidConfigurationException(msg.str());
+        }
+    }
+
+    // Evaluate the generator to produce a random delay.
+    Duration evaluate() {
+        auto value = number_generator->evaluate();
+        switch (units) {
+            case NANOSECOND:
+                return (std::chrono::duration_cast<Duration>(
+                    std::chrono::duration<double, std::nano>(value)));
+                break;
+            case MICROSECOND:
+                return (std::chrono::duration_cast<Duration>(
+                    std::chrono::duration<double, std::nano>(value)));
+                break;
+            case MILLISECOND:
+                return (std::chrono::duration_cast<Duration>(
+                    std::chrono::duration<double, std::milli>(value)));
+                break;
+            case SECOND:
+                return (std::chrono::duration_cast<Duration>(std::chrono::duration<double>(value)));
+                break;
+            case MINUTE:
+                return (std::chrono::duration_cast<Duration>(
+                    std::chrono::duration<double, std::ratio<60>>(value)));
+                break;
+            case HOUR:
+                return (std::chrono::duration_cast<Duration>(
+                    std::chrono::duration<double, std::ratio<3600>>(value)));
+                break;
+        }
+    }
+
+private:
+    UniqueGenerator<double> number_generator;
+    Units units;
+};
+
+struct Transition {
+    int next_state;
+    Delay delay;
+};
+
+// Represents one state in an FSM.
+struct State {
+public:
```

Comment:
💅🏽 in general prefer `class` to `struct` but that's no biggie. Either way, no need to put `public` in `struct`s 😸